src/main/java/org/truffleruby/core/hash/library/CompactHashStore.java

```diff
@@ -0,0 +1,666 @@
+            @Cached @Shared HashingNodes.ToHash hashFunction,
+            @Cached @Exclusive InlinedConditionProfile keyNotFound,
+            @Bind("$node") Node node) {
+        int keyHash = hashFunction.execute(key, hash.compareByIdentity);
+        long hashPosAndKeyKvPos = getHashPosAndKvPos.execute(key, keyHash, hash.compareByIdentity, index, kvStore);
+        int hashPos = IntPair.first(hashPosAndKeyKvPos);
+        int keyKvPos = IntPair.second(hashPosAndKeyKvPos);
+
+        if (keyNotFound.profile(node, keyKvPos == KEY_NOT_FOUND)) {
+            return null;
+        }
+
+        return deleteKvAndGetV(hash, hashPos, keyKvPos);
+    }
+
+    @ExportMessage
+    protected Object deleteLast(RubyHash hash, Object key,
+            @Cached @Shared HashingNodes.ToHash hashFunction,
+            @Cached @Shared GetHashPosForKeyAtKvPosNode getHashPos,
+            @Cached @Exclusive InlinedConditionProfile kvAllNulls,
+            @Cached @Exclusive InlinedConditionProfile keyGivenIsNotLastKey,
+            @Cached @Exclusive InlinedLoopConditionProfile nonNullKeyNotYetFound,
+            @Bind("$node") Node node) {
+        int lastKeyPos = firstNonNullKeyPosFromEnd(nonNullKeyNotYetFound, node);
+        if (kvAllNulls.profile(node, lastKeyPos == KEY_NOT_FOUND)) {
+            return null;
+        }
+        if (keyGivenIsNotLastKey.profile(node, key != kvStore[lastKeyPos])) {
+            return null;
+        }
+
+        int keyHash = hashFunction.execute(key, hash.compareByIdentity);
+        int hashPos = getHashPos.execute(keyHash, lastKeyPos, index);
+
+        return deleteKvAndGetV(hash, hashPos, lastKeyPos);
+    }
+
+    @ExportMessage
+    protected Object eachEntry(RubyHash hash, EachEntryCallback callback, Object state,
+            @Cached @Exclusive InlinedConditionProfile keyNotNull,
+            @Cached @Exclusive InlinedLoopConditionProfile tillArrayEnd,
+            @Bind("$node") Node node) {
+        for (int i = 0; tillArrayEnd.inject(node, i < kvStore.length); i += 2) {
+            if (keyNotNull.profile(node, kvStore[i] != null)) {
+                callback.accept(i / 2, kvStore[i], kvStore[i + 1], state);
+            }
+        }
+        return state;
+    }
+
+    @ExportMessage
+    protected Object eachEntrySafe(RubyHash hash, EachEntryCallback callback, Object state,
+            @CachedLibrary("this") HashStoreLibrary hashlib) {
+        return hashlib.eachEntry(this, hash, callback, state);
+    }
+
+    @ExportMessage
+    protected void replace(RubyHash hash, RubyHash dest,
+            @Cached @Exclusive PropagateSharingNode propagateSharing,
+            @Cached @Exclusive InlinedConditionProfile noReplaceNeeded,
+            @Bind("$node") Node node) {
+        if (noReplaceNeeded.profile(node, hash == dest)) {
+            return;
+        }
+
+        propagateSharing.execute(node, dest, hash);
+
+        CompactHashStore copy = this.copy();
```

Comment:
This ordering won't share the keys/values from `dest`, if `hash` is shared, which is incorrect.
We should add a spec for that case if there isn't one yet.

Just swapping the order will not work either, because WriteBarrierNode on `hash` if it is already shared will noop and not share the keys/values.
So we need to share the keys/values explicitly here, unless `dest` is shared in which case there is nothing to do.

Reply:
Ah actually I think it's fine.
If `hash` is shared, then its elements are already shared, and we only need to mark `dest` itself as shared.
If `hash` is not shared, nothing to do.